### PR TITLE
feat: add application.UseEmailAsSamlNameId field for SAML

### PR DESCRIPTION
**Pull request by okatu-loli**
Fix: #3077

### CODE REVIEW

File: object/saml_idp.go [hunk 1]
```diff
@@ -386,7 +386,7 @@ func GetSamlResponse(application *Application, user *User, samlRequest string, h
 }
 
 // NewSamlResponse11 return a saml1.1 response(not 2.0)
-func NewSamlResponse11(user *User, requestID string, host string) (*etree.Element, error) {
+func NewSamlResponse11(application *Application, user *User, requestID string, host string) (*etree.Element, error) {
```
Comment (dacongda) [inline object/saml_idp.go:389]: i think we should also change `NewSamlResponse` in line 40

File: object/application.go [hunk 2]
```diff
@@ -78,6 +78,7 @@ type Application struct {
 	EnableSamlCompress    bool            `json:"enableSamlCompress"`
 	EnableSamlC14n10      bool            `json:"enableSamlC14n10"`
 	EnableSamlPostBinding bool            `json:"enableSamlPostBinding"`
+	UseEmailAsNameId      bool            `json:"useEmailAsNameId"`
```
Comment (dacongda) [inline object/application.go:81]: Maybe `UseEmailAsSamlNameId`  is a better expression

File: object/saml_idp.go [hunk 3]
```diff
@@ -439,7 +439,11 @@ func NewSamlResponse11(user *User, requestID string, host string) (*etree.Elemen
 	attributeStatement := assertion.CreateElement("saml:AttributeStatement")
 	subjectInAttribute := attributeStatement.CreateElement("saml:Subject")
 	nameIdentifierInAttribute := subjectInAttribute.CreateElement("saml:NameIdentifier")
-	nameIdentifierInAttribute.SetText(user.Name)
+	if application.UseEmailAsNameId {
+		nameIdentifierInAttribute.SetText(user.Email)
```
Comment (dacongda) [inline object/saml_idp.go:443]: why not change `nameIdentifier` in line 433

**Comment by dacongda** [comment]
can you record a demo video or show different in response xml？

**Comment by okatu-loli** [comment]
> can you record a demo video or show different in response xml？

OK, I'll reply later

**Comment by okatu-loli** [comment]
> can you record a demo video or show different in response xml？

image

**Review by dacongda — APPROVED** [review]

**Pull request merged**